stint/io.nim

```diff
@@ -104,87 +126,67 @@ func stuint*(a: StUint, bits: static[int]): StUint[bits] {.inline.} =
-#   ## signed int to unsigned int conversion
-#   ## current behavior is cast-like, copying bit pattern
-#   ## or truncating if input does not fit into destination
-#   const N = bitsof(x.data)
-#   when N < bits:
-#     when N <= 64:
-#       type T = StUint[N]
-#       result = StUint(convert[T](a).data, bits)
-#     else:
-#       smallToBig(result.data, a.data)
-#   elif N > bits:
-#     when bits <= 64:
-#       result = StUint(x.truncate(type(result.data)), bits)
-#     else:
-#       bigToSmall(result.data, a.data)
-#   else:
-#     result = convert[type(result)](a)
-
-# func stint*(a: StInt, bits: static[int]): StInt[bits] {.inline.} =
-#   ## signed int to signed int conversion
-#   ## will raise exception if input does not fit into destination
-#   const N = bitsof(a.data)
-#   when N < bits:
-#     when N <= 64:
-#       result = stint(a.data, bits)
-#     else:
-#       if a.isNegative:
-#         smallToBig(result.data, (-a).data)
-#         result = -result
-#       else:
-#         smallToBig(result.data, a.data)
-#   elif N > bits:
-#     template checkNegativeRange() =
-#       # due to bug #92, we skip negative range check
-#       when false:
-#         const dmin = stint((type result).low, N)
-#         if a < dmin: raise newException(RangeError, "value out of range")
-
-#     template checkPositiveRange() =
-#       const dmax = stint((type result).high, N)
-#       if a > dmax: raise newException(RangeError, "value out of range")
-
-#     when bits <= 64:
-#       if a.isNegative:
-#         checkNegativeRange()
-#         result = stint((-a).truncate(type(result.data)), bits)
-#         result = -result
-#       else:
-#         checkPositiveRange()
-#         result = stint(a.truncate(type(result.data)), bits)
-#     else:
-#       if a.isNegative:
-#         checkNegativeRange()
-#         bigToSmall(result.data, (-a).data)
-#         result = -result
-#       else:
-#         checkPositiveRange()
-#         bigToSmall(result.data, a.data)
-#   else:
-#     result = a
-
-# func stint*(a: StUint, bits: static[int]): StInt[bits] {.inline.} =
-#   const N = bitsof(a.data)
-#   const dmax = StUint((type result).high, N)
-#   if a > dmax: raise newException(RangeError, "value out of range")
-#   when N < bits:
-#     when N <= 64:
-#       result = stint(a.data, bits)
-#     else:
-#       smallToBig(result.data, a.data)
-#   elif N > bits:
-#     when bits <= 64:
-#       result = stint(a.truncate(type(result.data)), bits)
-#     else:
-#       bigToSmall(result.data, a.data)
-#   else:
-#     result = convert[type(result)](a)
+  when bits <= a.bits:
+    for i in 0 ..< result.len:
+      result[i] = a[i]
+  else:
+    for i in 0 ..< a.len:
+      result[i] = a[i]
+
+func stuint*(a: StInt, bits: static[int]): StUint[bits] {.inline.} =
+  ## signed int to unsigned int conversion
+  ## bigger to smaller bits conversion, the result is truncated
+  doAssert(a.isPositive, "Cannot convert negative number to unsigned int")
```

Comment:
``OverflowDefect`` is supposed to handle Overflow and Underflow
or ``RangeDefect``

Reply:
Fixed